lib/Parse/ParseDecl.cpp

```diff
@@ -261,92 +259,135 @@ void Parser::parseTopLevelItems(SmallVectorImpl<ASTNode> &items) {
+      // SwiftParser
+      SourceLoc loc;
+      if (auto bufferID = SF.getBufferID()) {
           loc = Context.SourceMgr.getLocForBufferStart(*bufferID);
-        }
-        diagnose(loc, diag::parser_new_parser_errors);
       }
+      diagnose(loc, diag::parser_new_parser_errors);
     }
   }
 #endif
 }
 
+void *ExportedSourceFileRequest::evaluate(Evaluator &evaluator,
+                                          const SourceFile *SF) const {
+  // The SwiftSyntax parser doesn't (yet?) handle SIL.
+  if (SF->Kind == SourceFileKind::SIL)
+    return nullptr;
+
+#if SWIFT_SWIFT_PARSER
+  auto &ctx = SF->getASTContext();
+  auto &SM = ctx.SourceMgr;
+
+  auto bufferID = SF->getBufferID();
+  if (!bufferID)
+    return nullptr;
+
+  StringRef contents = SM.extractText(SM.getRangeForBuffer(*bufferID));
+
+  // Parse the source file.
+  auto exportedSourceFile = swift_ASTGen_parseSourceFile(
+      contents.begin(), contents.size(),
+      SF->getParentModule()->getName().str().str().c_str(),
+      SF->getFilename().str().c_str());
+
+  ctx.addCleanup([exportedSourceFile] {
+    swift_ASTGen_destroySourceFile(exportedSourceFile);
+  });
+  return exportedSourceFile;
+#else
+  return nullptr;
+#endif
+}
+
 void
 Parser::parseSourceFileViaASTGen(SmallVectorImpl<ASTNode> &items,
                                  Optional<DiagnosticTransaction> &transaction,
                                  bool suppressDiagnostics) {
 #if SWIFT_SWIFT_PARSER
-  Optional<DiagnosticTransaction> existingParsingTransaction;
-  if (SF.Kind != SourceFileKind::SIL) {
-    StringRef contents =
-        SourceMgr.extractText(SourceMgr.getRangeForBuffer(L->getBufferID()));
-
-    // Parse the source file.
-    auto exportedSourceFile = swift_ASTGen_parseSourceFile(
-        contents.begin(), contents.size(),
-        SF.getParentModule()->getName().str().str().c_str(),
-        SF.getFilename().str().c_str());
-    SF.exportedSourceFile = exportedSourceFile;
-    Context.addCleanup([exportedSourceFile] {
-      swift_ASTGen_destroySourceFile(exportedSourceFile);
-    });
+  using ParsingFlags = SourceFile::ParsingFlags;
+  const auto parsingOpts = SF.getParsingOptions();
 
-    // If we're supposed to emit diagnostics from the parser, do so now.
-    if ((Context.LangOpts.hasFeature(Feature::ParserDiagnostics) ||
-         Context.LangOpts.hasFeature(Feature::ParserASTGen)) &&
-        !suppressDiagnostics &&
-        swift_ASTGen_emitParserDiagnostics(
-            &Context.Diags, SF.exportedSourceFile, /*emitOnlyErrors=*/false,
-            /*downgradePlaceholderErrorsToWarnings=*/
-                Context.LangOpts.Playground ||
-                Context.LangOpts.WarnOnEditorPlaceholder) &&
-        Context.Diags.hadAnyError() &&
-        !Context.LangOpts.hasFeature(Feature::ParserASTGen)) {
+  // We only need to do parsing if we either have ASTGen enabled, or want the
+  // new parser diagnostics.
+  auto needToParse = [&]() {
```

Comment:
Was this to avoid negating the if's here 😅?

Reply:
Yeah I originally wrote it as:

```cxx
    if (!parsingOpts.contains(ParsingFlags::EnableASTGen) &&
        !(parsingOpts.contains(ParsingFlags::EnableASTGenDiagnostics) && !suppressDiagnostics))
      return;
```

and then I said to myself: "I'd hate to read that code if I ever saw it in the wild", and so rewrote it as this 😅

Reply:
You could split it across two if and just negate with an or instead:
```
if (!parsingOpts.contains(ParsingFlags::EnableASTGen)) {
  if (!parsingOpts.contains(ParsingFlags::EnableASTGenDiagnostics) || suppressDiagnostics)
    return
}
```
I don't have a strong preference. I think yours is probably clearer.

Reply:
Hmm I kind of also dislike that too